File: game/forest/win_talking.py
import pygame

from forest.forest_resources import ForestResources
from null_state import NullState
from state import State


class WinTalking(State):
    def process_events(self, events):
        if self.get_frame_index() >= len(ForestResources.sync_levelcompleted):
            return NullState(self.parent)

    def on_enter(self) -> None:
        super().on_enter()
        pygame.mixer.Sound.play(ForestResources.speak_levelcompleted)

    def render(self, screen):
        self.parent.render_background(screen)
        self.parent.render_bottom(screen)
        screen.blit(ForestResources.hugo_telllives[ForestResources.sync_levelcompleted[self.get_frame_index()]], (128, -16))

File: game-web/src/forest/winTalking.ts
/** Port of game/forest/win_talking.py */

import { State, type StateClass } from '../state';
import { Animation } from '../animation';
import { AudioHelper } from '../audioHelper';
import { ForestResources } from './forestResources';
import { NullState } from '../nullState';
import { RenderType } from '../renderType';
import type { GameData } from '../gameData';
import type { PhoneEvents } from '../phoneEvents';

export class WinTalking extends State {
  readonly needsBackground = RenderType.PRE;
  readonly needsBottom = RenderType.PRE;

  constructor(context: GameData) { super(context); }

  processEvents(_events: PhoneEvents): StateClass | null {
    return this.getFrameIndex() >= ForestResources.syncLevelcompleted.length ? NullState : null;
  }

  onEnter(): void {
    super.onEnter();
    this.context.forestReachedEnd = true;
    AudioHelper.play(ForestResources.speakLevelcompleted);
  }

  render(ctx: OffscreenCanvasRenderingContext2D): void {
    ctx.drawImage(Animation.getSyncFrame(ForestResources.hugoTelllives, ForestResources.syncLevelcompleted, this.getFrameIndex()), 128, -16);
  }
}
